Prefer equality in boolean comparisons

ranma42 opened the pull request:
In most databases, equality comparisons can take advantage of indexing and inequalities cannot.

Fixes #34164.

ranma42 committed:
Prefer equality in boolean comparisons

They can take advantage of indexing.

ranma42 committed:
Update baselines

roji commented on test/EFCore.Sqlite.FunctionalTests/Query/NullSemanticsQuerySqliteTest.cs:
```diff
@@ -839,7 +839,7 @@ public override async Task Rewrite_compare_bool_with_bool(bool async)
 SELECT "e"."Id"
 FROM "Entities1" AS "e"
-WHERE "e"."BoolA" <> "e"."NullableBoolB"
+WHERE "e"."BoolA" = (NOT ("e"."NullableBoolB"))
```
Is this a positive change? I doubt any database out there will use index with equality and NOT, more than it would for inequality, no? Should we make this change more targeted, so that it doesn't do this specific transformation?

ranma42 replied:
yes, it is, at least as long as we do not analyze the provenance of the columns.
In this specific case, it is basically irrelevant: the whole table is going to be scanned linearly (just once).
If the left column and the right column came from different tables, it would be much more efficient.

This is basically an instance of https://github.com/dotnet/efcore/pull/34048#discussion_r1661103828

ranma42 replied:
I'll add some examples in the issue #34164.

ranma42 replied:
Added Sqlite and Postgres examples 🚀

roji replied:
So between `a <> b` and `a = NOT(b)`, the former certainly seems  more natural, and what I'd expect a standard SQL query to look like. If these perform the same, I'd definitely prefer the first, at least for readability etc. (and we do generally care about that).

I could alsi imagine a database where the planner optimizes to use an index on b (effectively transforming the inequality into a not on a), where this change would cause a regression. of course, this is entirely speculative, and I have an actually looked into which databases basis to which optimizations.

I'd prefer us to do a bit more cross database research before merging a change like this, which at the very least makes our SQL less readable/standard/expected (and that does tend to have some correlation sometimes with performance). If this address is a very specific Sqlite behavior, where equality is always better, we always have the option of doing this change for sqlite only.

roji replied:
OK, I wrote the above comment before noticing you posted data on other databases… some remarks:

* For the Sqlite case, have you confirmed that the second option, where an index is built, is actually faster than the first?
* For the SQL Server case, the total subtree cost is actually higher with the second method.

ranma42 replied:
I will perform some measurement, but I do not have real code/an actual database that is using this filter; I will try to do some synthetic examples (I'll try to cover some interesting cases, but a real-world case would definitely be more relevant).

ranma42 replied:
Some (synthetic) benchmarks have been posted in https://github.com/dotnet/efcore/issues/34164

ranma42 commented:
Maybe I should have explained this in advance: this is not a new/different approach to translate boolean (in)equalities; it is just making the code more consistent in choosing `=` over `<>` (as per the issue).

Ideally there should be no duplication of this code, hence it should be "inevitably" consistent.

ranma42 commented:
I cleaned up the code a little (the same equality conversion logic is now shared between `OptimizeComparison` and `RewriteNullSemantics`.

ranma42 commented:
Rebased to resolve conflicts

ranma42 commented:
Rebased to resolve conflicts

maumar committed:
Fix to #35393 - GroupJoin in EF Core 9 Returns Null for Joined Entities

Problem was that in EF9 we moved some optimizations from sql nullability processor to SqlExpressionFactory (so that we optimize things early). One of the optimizations:

```
!(true == a) -> false == a
!(false == a) -> true == a
```

is not safe to do when a is a constant or parameter null value, because it evaluates to true IS NULL (two value logic). Fix is to remove this optimization from SqlExpressionFactory and instead put it back into OptimizeNotExpression, once we've converted everything we could to IS NULL checks already.

Fixes #35393

ranma42 commented on src/EFCore.Relational/Query/SqlNullabilityProcessor.cs:
```diff
@@ -1589,22 +1565,49 @@ private SqlExpression RewriteNullSemantics(
+        var notEqual = sqlBinaryExpression.OperatorType == ExpressionType.Equal ^ leftNegated == rightNegated;
+
+        // prefer equality in predicates
+        if (optimize && notEqual && left.Type == typeof(bool))
```
this should be modified to take into account value converters

ranma42 replied:
not relevant

ranma42 committed:
Update baselines

These are the baselines regenerated after a rebase on top of
35fc423ba39630b3bf883e9e06c4ee4ea038ae3e.

ranma42 commented:
This is currently marked as draft as it is based on (an old version of):
 - https://github.com/dotnet/efcore/pull/34166

maumar commented:
/azp run

maumar commented:
@ranma42 I think it makes sense to use this PR in favor of [mine ](https://github.com/dotnet/efcore/pull/35395) as the basis for the fix, at least for main. We could consider scoping it a bit for patch, but that can be a separate discussion

ranma42 commented:
👍 @maumar your fix is still needed, at least the part that drops the invalid optimization from `SqlExpressionFactory`, but we can probably avoid re-introducing the code in `SqlNullabilityProcessor`.
I will work on cleaning this up a bit and experiment with showing & fixing the regression that occurs on value-converted types as mentioned in https://github.com/dotnet/efcore/pull/35395#discussion_r1900377414

maumar commented:
@ranma42 also consider extending Rewrite_compare_bool_with_bool test to include constants. I did that as part of the patch fix and it indeed caught the error https://github.com/dotnet/efcore/commit/e12547fb026af02fa88dee655da4770673f68fac
